hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -105,6 +107,49 @@ else if ( uniqueSemantic == UniqueSemantic.ASSERT ) {
 		}
 	}
 
+	private static class Results<R> {
+		private final List<R> results = new ArrayList<>();
+		private final JavaType resultJavaType;
+
+		public Results(JavaType resultJavaType) {
+			this.resultJavaType = resultJavaType;
+		}
+
+		public boolean contains(R result) {
+			for ( int i = 0; i < results.size(); i++ ) {
+				if ( resultJavaType.areEqual( results.get( i ), result ) ) {
+					return true;
+				}
+			}
+			return false;
+		}
```

Comment:
I think I'd prefer having a `boolean addUnique(R result)` instead which returns whether the object was added or not. Enables implementing it as `if ( added.add( result ) ) { getResults().add( result ); }` instead which is a bit faster.